subprojects/docs/src/docs/userguide/authoring-builds/basics/writing_tasks.adoc

```diff
@@ -0,0 +1,189 @@
+    fun action() {
+        val file = File("myfile.txt")
+        file.createNewFile()
+        file.writeText("HELLO FROM MY TASK")
+    }
+}
+----
+
+The `CreateFileTask` implements a simple set of *actions*.
+First, a file called "myfile.txt" is created in the main project.
+Then, some text is written to the file.
+
+== Register a task
+
+A task is *registered* in the build script using the link:{javadocPath}/org/gradle/api/tasks/TaskContainer.html[`TaskContainer.register()`] method, which allows it to be then used in the build logic.
+
+.build.gradle.kts
+[source,kotlin]
+----
+abstract class CreateFileTask : DefaultTask() {
+    @TaskAction
+    fun action() {
+        val file = File("myfile.txt")
+        file.createNewFile()
+        file.writeText("HELLO FROM MY TASK")
+    }
+}
+
+tasks.register<CreateFileTask>("createFileTask") {
+
+}
+----
+
+== Task group and description
+
+Setting the *group* and *description* properties on your tasks can help users understand how to use your task:
+
+.build.gradle.kts
+[source,kotlin]
+----
+abstract class CreateFileTask : DefaultTask() {
+    @TaskAction
+    fun action() {
+        val file = File("myfile.txt")
+        file.createNewFile()
+        file.writeText("HELLO FROM MY TASK")
+    }
+}
+
+tasks.register<CreateFileTask>("createFileTask", ) {
+    group = "custom"
+    description = "Create myfile.txt in the current directory"
+}
+----
+
+Once a task is added to a group, it is visible when listing tasks.
+
+== Task input and outputs
+
+For the task to do useful work, it typically needs some *inputs*.
+A task typically produces *outputs*.
+
+.build.gradle.kts
+[source,kotlin]
+----
+abstract class CreateFileTask : DefaultTask() {
+    @Input
+    val fileText = "HELLO FROM MY TASK"
+
+    @Input
+    val fileName = "myfile.txt"
+
+    @OutputFile
+    val myFile: File = File(fileName)
+
+    @TaskAction
+    fun action() {
+        myFile.createNewFile()
+        myFile.writeText(fileText)
+    }
+}
+
+tasks.register<CreateFileTask>("createFileTask") {
+    group = "custom"
+    description = "Create myfile.txt in the current directory"
+}
+----
+
+== Configure a task
+
+A task is optionally *configured* in a build script using the link:{javadocPath}/org/gradle/api/tasks/TaskCollection.html[`TaskCollection.named()`] method.
+
+The `CreateFileTask` class is updated so that the text in the file is configurable:
+
+.build.gradle.kts
+[source,kotlin]
+----
+abstract class CreateFileTask : DefaultTask() {
+    @get:Input
+    abstract val fileText: Property<String>
+
+    @Input
+    val fileName = "myfile.txt"
+
+    @OutputFile
+    val myFile: File = File(fileName)
+
+    @TaskAction
+    fun action() {
+        myFile.createNewFile()
+        myFile.writeText(fileText.get())
+    }
+}
+
+tasks.register<CreateFileTask>("createFileTask") {
+    group = "custom"
+    description = "Create myfile.txt in the current directory"
+}
+
+tasks.named<CreateFileTask>("createFileTask") {
```

Comment:
I wanted to use a single complete example for this one. It's unfortunately not very realistic. I couldn't think of anything better. Open to suggestions but would like to avoid a built-in task in that section.

Reply:
You could double-down on using Properties:

```kotlin
tasks.register<CreateFileTask>("createFileTask") {
    group = "custom"
    description = "Create myfile.txt in the current directory"
	fileText.convention("HELLO FROM THE CREATE FILE TASK METHOD")
}

tasks.named<CreateFileTask>("createFileTask") {
	fileText.set("HELLO TO ALL, AND HAVE A WONDERFUL DAY!")
}
```

But I think that only works if you've established how Properties work first.  If you change to just a `String` field you could append to the `fileText` without distracting `Property` get/set calls.  I'm not sure what's best here.  Maybe new readers won't even notice this.